### PR TITLE
implementation of csep1 tests

### PR DESCRIPTION
**Pull request by khawajasim**
Hi Bill,

Update: New commit (18-2-20). I have also added implementation of T test and W test to the code. I request someone to please look into the mathematics of tests. 

update: Commit Request

I created codes of L,M,S,N Tests along with their unit tests and plotting utility. 
Addressed some of comments.
Please have a look at them and if found suitable, please merge them. 

Cheers
Asim

### CODE REVIEW

**Comment by wsavran** [comment]
Great I'll post any comments from the code review here

File: tests/csep1_tests.py [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+        forecast: Actual Forecast of a Model (Grided)
+    
+    Returns:
+        A new forecast which is the realization of Actual forecast
+    """
+        
+    return scipy.stats.poisson.ppf(random_matrix, forecast) 
+
+
+def l_test(observation, forecast, num_simulations, model_name):
```
Comment (wsavran) [inline tests/csep1_tests.py]: this doesn't look to be the optimized likelihood function. this is ok for now, but we will want to change that

File: tests/csep1_tests.py [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+        # computes the inverse cdf of the poisson random variable
+        forecast_realization = forecast_realization_inverse_cdf(random_matrix, forecast)  #Wrapper is put around scipy inverse poisson cdf
+        #2) Apply 2nd Test of FORECAST REALIZATION by Feeding CHOSEN RANDOM MATRIX
+        #print(forecast_realization)
+        # joint-loglikelihood from forecast realization and offset to zero for simplifed quantile score calc. 
+        #print(forecast_realization)
+        ll_simulations[count] = numpy.sum(poisson_log_likelihood(forecast_realization, forecast))
+        #print(ll_simulations)
+    # compute quantile score from Eq. 15 from Zechar et al., 2010    
+    quantile_score = numpy.sum(ll_simulations <= ll_actual) / num_simulations #Apply 3rd UNIT TEST here on Quantile Score.
```
Comment (wsavran) [inline tests/csep1_tests.py]: re: comment

File: csep/utils/csep1_unit_tests [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        print('L Test: Step 1, Running NUMPY Test')
+        numpy.testing.assert_array_equal(poisson_log_likelihood(observation, forecast), Actual_log_Likelihood,'Failed L Test: Actual Likelihood Testing Failed')
+        
+        #Step 2: Test Forecast Realization. 
+        print('L Test: Step 2, Running NUMPY Test')
+        numpy.testing.assert_array_equal(forecast_realization_inverse_cdf(random_matrix, forecast), forecast_realization, 'Failed L Test: Simulated Forecast Likelihood Test')
+        
+        #Step 3: Check Equation of Quantile Score
+        print('L Test: Step 3, Running Unit Test')
+        self.assertEqual((numpy.sum(ll_simulations <= ll_actual) / num_simulations), Quantile_score,'Failed L Test: Quantile Score Testing')
```
[the remark this answers is not in the document]
Comment (khawajasim) [reply csep/utils/csep1_unit_tests]: sum(ll_simulations <= ll_actual) / num_simulations,
is the equation (15) of Zechar et al. 2010. We intend to count all those Log Likelihoods of simulated forecasts, that are not greater than Actual Log likelihood. There this sign "<=" is employed to do the job.

File: csep/utils/csep1_unit_tests [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        forecast = numpy.array([[8, 2],[3,5]])
+        
+        observation_magbins = numpy.array([9, 10])
+        norm_forecast_magbins = numpy.array([11, 7])*19/18
+        
+        #Apply Test now
+        print('M Test: Observation Preparation, Running NUMPY Test')
+        numpy.testing.assert_array_equal(m_test_prep(observation,forecast)[0],observation_magbins,'M Test: Failed Row Summation of Observation')
+        print('M Test: Forecast Preparation, Running NUMPY Test')
+        numpy.testing.assert_array_equal(m_test_prep(observation,forecast)[1],norm_forecast_magbins,'M Test: Failed Row Summation or Normalization of Forecast')
```
[the remark this answers is not in the document]
Comment (khawajasim) [reply csep/utils/csep1_unit_tests]: Thank you for suggestion. I shall incorporate these suggestions in a few days, after dealing with other tests.
The same response for other suggested changes. 
Regards

File: csep/utils/plotting.py [hunk 5]
```diff
@@ -950,4 +950,183 @@ def plot_probability_test(evaluation_result, axes=None, plot_args=None, show=Tru
...
+#observation = numpy.zeros((3,3))+8
+#L2 = m_test(observation, forecast, 5,'M')
+#observation=numpy.array([[5, 8],[4, 2]])
+#forecast = numpy.array([[8, 2],[3, 5]])
+#L3 = l_test(observation, forecast, 5,'L')
+#Model_list = [L1,L2,L3]
+#multiple_ll_models(Model_list)
+#single_ll_model(L1)
+#single_ll_model(L2)
+#single_ll_model(L3)
```
[the remark this answers is not in the document]
Comment (khawajasim) [reply csep/utils/plotting.py]: Got that. I would delete these comments.

File: tests/csep1_tests.py [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+        observation: Observed (Grided) seismicity
+        forecast: Forecast of a Model (Grided)
+    
+    Returns:
+        Log-Liklihood values of between binned-observations and binned-forecasts
+    """
+    return numpy.log(scipy.stats.poisson.pmf(observation, forecast))
+
+
+def forecast_realization_inverse_cdf(random_matrix, forecast):
```
[the remark this answers is not in the document]
Comment (khawajasim) [reply tests/csep1_tests.py]: Understood

File: tests/csep1_tests.py [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+        raise TypeError('observation must be numpy.ndarray')
+    if not isinstance(forecast, numpy.ndarray):
+        raise TypeError('forecast must be numpy.ndarray')
+        
+    # seed the random number generator
+    #new_seed = seed or numpy.random.randint(1, high=99999)
+    #numpy.random.seed(seed=new_seed)
+    
+    # compute log-likelihood between the forecast and observations
+    ll_actual = numpy.sum(poisson_log_likelihood(observation, forecast))   #1) Apply first UNIT Test Here BEFORE NUMPYSUM...
```
[the remark this answers is not in the document]
Comment (khawajasim) [reply tests/csep1_tests.py]: Which comment? Are you referring about moving this function to csep.utils.stats ?

File: tests/csep1_tests.py [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+    #print(ll_actual)
+    # pre-allocated vector to store simulated log-likelihoods
```
[the remark this answers is not in the document]
Comment (khawajasim) [reply tests/csep1_tests.py]: Understood

File: tests/csep1_tests.py [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+#    plt.axvline(x=ll_actual)
+#    plt.show()
+    L_Test_Eval = { 'model_name': model_name,
+                    'quantile' : quantile_score,
+                    'll_actual': ll_actual,
+                     'll_simulation':ll_simulations}
+     
+    return L_Test_Eval
+        
+def m_test_prep(observation, forecast):
```
[the remark this answers is not in the document]
Comment (khawajasim) [reply tests/csep1_tests.py]: understood

File: tests/csep1_tests.py [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+    observations_magbins = numpy.sum(observation, axis=0)
+    forecast_magbins = numpy.sum(forecast,axis=0)
+    
+    norm_forecast_magbins = forecast_magbins*N_obs/N_fcst
+    
+    return observations_magbins, norm_forecast_magbins
+
+
+
+def m_test(observation, forecast, num_simulations, model_name):
```
[the remark this answers is not in the document]
Comment (khawajasim) [reply tests/csep1_tests.py]: Understood

File: tests/csep1_tests.py [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+    if not isinstance(forecast, numpy.ndarray):
+        raise TypeError('forecast must be numpy.ndarray')
+    
+    [observations_magbins, norm_forecast_magbins] = m_test_prep(observation,forecast)
+  
+    M_Test_Eval = l_test(observations_magbins, norm_forecast_magbins, num_simulations, model_name)
+    return M_Test_Eval
+
+
+def s_test_prep(observation, forecast):
```
[the remark this answers is not in the document]
Comment (khawajasim) [reply tests/csep1_tests.py]: understood

File: tests/csep1_tests.py (hunk 1 above)
Comment (khawajasim) [reply tests/csep1_tests.py]: Actually, 
I looked into the implementation of the optimized version of L test, implemented in MATLAB. I could reproduce that easily, but to be honest that was not appealing to my understanding, therefore I did not implement that. Let me walk you through that piece of code here, then we can decide what to do about it. 

According to my understanding, that implementing is using two loops. Furthermore, the method by which forecasat realization is computed does not sound interesting at all. See the steps below and then see the corresponding piece of code below: 

Step 1- for loop explicitly used
Step 2- Using "find" function of Matlab to look for certain criterion
Step 3- Then computing the realization of forecast by just adding +1 value to the actual forecast (means, no inverse Poisson CDF)
Step 4-  Compute Likelihood of "actual_forecast_value+1" AND "actual_forecast_value". 
Step 5- After loops end, sum all likelihoods. 
----------------Code Lines-------------
vRandom = pythonCSEPRandom(nNumberEvents, 1); 
vRatesCDF = [cumsum(vForecast(:, Forecast_Rate)./fTotalForecastRate)];

for nCnt = 1:nNumberEvents     (step 1)
  nIndex = min(find((vRandom(nCnt) < vRatesCDF)));       (step 2)
  vForecast(nIndex, Forecast_Observations) = vForecast(nIndex, Forecast_Observations) + 1;  (step 3)
  vForecast(nIndex, Forecast_PrecomputedZeroLikelihood) = calc_logpoisspdf(vForecast(nIndex, Forecast_Observations), vForecast(nIndex, Forecast_Rate)); (   step 4)
end;
fLikelihood = sum(vForecast(:, Forecast_PrecomputedZeroLikelihood)); (step 5)
--------------Code Ends-----------------------
This does not make sense all to me. If I am wrong, please guide me. 

However, I have found another implementation of Likelihood:
    Rhoades et al. 2011. (Equation 4): 
             A revised equation of LogLikelihood formula is provided. Which I can be a bit efficient, in a way that it skips the use of a Log and a Factorial. Rest of all steps to be remain same. 

Please enlighten me, if I am not getting it right.
Regards.

Comment (khawajasim) [reply tests/csep1_tests.py]: Actually, 
I looked into the implementation of the optimized version of L test, implemented in MATLAB. I could reproduce that easily, but to be honest that was not appealing to my understanding, therefore I did not implement that. Let me walk you through that piece of code here, then we can decide what to do about it. 

According to my understanding, that implementation is using two loops (so not much faster). Furthermore, the method by which forecast realization is computed does not appeal to my understanding. See the steps below and then match the corresponding piece of code below: 

Step 1- for loop explicitly used
Step 2- Using "find" function of Matlab to look for certain criterion
Step 3- Then computing the realization of forecast by just adding +1 value to the actual forecast (means, no inverse Poisson CDF)
Step 4-  Compute Likelihood of "actual_forecast_rate+1" AND "actual_forecast_rate". 
Step 5- After loops end, sum all likelihoods. 
----------------Code Lines-------------
vRandom = pythonCSEPRandom(nNumberEvents, 1); 
vRatesCDF = [cumsum(vForecast(:, Forecast_Rate)./fTotalForecastRate)];

for nCnt = 1:nNumberEvents     (step 1)
  nIndex = min(find((vRandom(nCnt) < vRatesCDF)));       (step 2)
  vForecast(nIndex, Forecast_Observations) = vForecast(nIndex, Forecast_Observations) + 1;  (step 3)
  vForecast(nIndex, Forecast_PrecomputedZeroLikelihood) = calc_logpoisspdf(vForecast(nIndex, Forecast_Observations), vForecast(nIndex, Forecast_Rate)); (   step 4)
end;
fLikelihood = sum(vForecast(:, Forecast_PrecomputedZeroLikelihood)); (step 5)
--------------Code Ends-----------------------
This does not make sense all to me. If I am wrong, please guide me. 

However, I have found another implementation of Likelihood:
    Rhoades et al. 2011. (Equation 4): 
             A revised equation of LogLikelihood formula is provided. Which I think can be a bit efficient, in a way that it skips the use of a Log and a Factorial. Rest of all steps to be remained same. 

Please enlighten me, if I am not getting it right.
Regards.

File: csep/utils/plotting.py [hunk 12]
```diff
@@ -950,4 +950,183 @@ def plot_probability_test(evaluation_result, axes=None, plot_args=None, show=Tru
     if show:
         pyplot.show()
 
-    return ax
\ No newline at end of file
+    return ax
+
+##---Plotting functions for Likelihood based tests and N test, By Asim
+
+def multiple_ll_models(Model_list):
```
[the remark this answers is not in the document]
Comment (khawajasim) [reply csep/utils/plotting.py]: Skipping This comment for the end. As I might need to discuss it to fully understand your point.

File: csep/utils/plotting.py [hunk 13]
```diff
@@ -950,4 +950,183 @@ def plot_probability_test(evaluation_result, axes=None, plot_args=None, show=Tru
...
+    
+    ########## - Just Actual Likelihoods
+    fig, ax2=plt.subplots()
+    ax2.scatter(name,ll_actuals)   #(name,quantile)
+    plt.xlabel('Model Names')
+    plt.ylabel('Joint Log Likelihood')
+    plt.title('Joint Log Likelihood between actual forecast and observation')
+    
+
+def single_ll_model(Single_ll_Model):
```
[the remark this answers is not in the document]
Comment (khawajasim) [reply csep/utils/plotting.py]: Skipping This comment for the end. As I might need to discuss it to fully understand your point.

File: csep/utils/csep1_unit_tests [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+import unittest
+from csep1_tests import n_test
+from csep1_tests import poisson_log_likelihood
+from csep1_tests import _forecast_realization_inverse_cdf
+from csep1_tests import _m_test_prep
+from csep1_tests import _s_test_prep
+from csep1_tests import _conditional_l_test_prep
+
+
+class TestGriddedForecastTests(unittest.TestCase):
```
Comment (wsavran) [inline csep/utils/csep1_unit_tests]: separate class for each test type or you can break up the unit tests so they only test a single aspect of the function.

File: csep/utils/csep1_unit_tests [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+        observation_magbins = numpy.array([9, 10])
+        norm_forecast_magbins = numpy.array([11, 7])*19/18
+        
+        print('M Test: Observation Preparation, Running NUMPY Test')
+        numpy.testing.assert_array_equal(_m_test_prep(observation,forecast)[0],observation_magbins,'M Test: Failed Row Summation of Observation')
+        print('M Test: Forecast Preparation, Running NUMPY Test')
+        numpy.testing.assert_array_equal(_m_test_prep(observation,forecast)[1],norm_forecast_magbins,'M Test: Failed Row Summation or Normalization of Forecast')
+        
+        
+    def test_s_test_partial(self):
```
Comment (wsavran) [inline csep/utils/csep1_unit_tests]: consider naming test_s_test_prep

File: tests/csep1_tests.py [hunk 16]
```diff
@@ -0,0 +1,336 @@
...
+    """
+                                                #, seed=None
+    if not isinstance(observation, numpy.ndarray):
+        raise TypeError('observation must be numpy.ndarray')
+    if not isinstance(forecast, numpy.ndarray):
+        raise TypeError('forecast must be numpy.ndarray')
+    
+    [observations_magbins, norm_forecast_magbins] = _m_test_prep(observation,forecast)
+  
+    M_Test_Eval = l_test(observations_magbins, norm_forecast_magbins, num_simulations, model_name)
```
Comment (wsavran) [inline tests/csep1_tests.py]: just add comment here to explain why we are using the l-test in the m-test

File: tests/csep1_tests.py [hunk 17]
```diff
@@ -0,0 +1,336 @@
...
+    ll_simulations = numpy.zeros(num_simulations)
+    
+    # compute simulated log-likelihoods
+    for count in range(num_simulations):
+        # random percentile for each independent bin
+        random_matrix = numpy.random.random(forecast.shape) 
+        # computes the inverse cdf of the poisson random variable
+        forecast_realization = _forecast_realization_inverse_cdf(random_matrix, forecast)  
+        ll_simulations[count] = numpy.sum(poisson_log_likelihood(forecast_realization, forecast))
+    quantile_score = numpy.sum(ll_simulations <= ll_actual) / num_simulations #Apply 3rd UNIT TEST here on Quantile Score.
```
Comment (wsavran) [inline tests/csep1_tests.py]: compute_quantile_score(...)

File: tests/csep1_tests.py (hunk 16 above)
Comment (wsavran) [inline tests/csep1_tests.py]: m_test_quantile = compute_quantile_score(observations_magbins, norm_forecast_magbins, num_simulations)

File: tests/csep1_tests.py [hunk 18]
```diff
@@ -0,0 +1,336 @@
...
+        # computes the inverse cdf of the poisson random variable
+        forecast_realization = _forecast_realization_inverse_cdf(random_matrix, forecast)  
+        ll_simulations[count] = numpy.sum(poisson_log_likelihood(forecast_realization, forecast))
+    quantile_score = numpy.sum(ll_simulations <= ll_actual) / num_simulations #Apply 3rd UNIT TEST here on Quantile Score. 
+    
+
+    L_Test_Eval = { 'model_name': model_name,
+                    'quantile' : quantile_score,
+                    'll_actual': ll_actual,
+                     'll_simulation':ll_simulations}
```
Comment (wsavran) [inline tests/csep1_tests.py]: we can use EvaluationResult

result = EvaluationResult(test_distribution=This is the ll_simulations
                          name=Name of forecast,
                          observed_statistic=<ll_actual>,
                          quantile=(delta_1, delta_2),
                          status='Normal',
                          obs_catalog_repr=obs.date_accessed,
                          sim_name=model_name,
                          min_mw=None,
                          obs_name=None)

File: csep/utils/csep1_unit_tests [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+        numpy.testing.assert_array_equal(poisson_log_likelihood(observation, forecast), Actual_log_Likelihood,'Failed L Test: Actual Likelihood Testing Failed')
+        
+        print('L Test: Step 2, Running NUMPY Test')
+        numpy.testing.assert_array_equal(_forecast_realization_inverse_cdf(random_matrix, forecast), forecast_realization, 'Failed L Test: Simulated Forecast Likelihood Test')
+        
+        print('L Test: Step 3, Running Unit Test')
+        self.assertEqual((numpy.sum(ll_simulations <= ll_actual) / num_simulations), Quantile_score,'Failed L Test: Quantile Score Testing')
+      
+        
+    def test_m_test_partial(self):
```
Comment (wsavran) [inline csep/utils/csep1_unit_tests]: im not sure these tests actually test the function itself. we want to call the function with the inputs and check that the outputs are what is expected.

these tests are looking to make sure indivdual parts of the function work but not the overall component.

Comment (khawajasim) [reply csep/utils/csep1_unit_tests]: Yes, you are right. When you were here in Potsdam and we started writing likelihood based tests (L test at that time) and we discussed this thing. Likelihood based tests use some random numbers generation inside them for creating forecasts realizations. So there were two options for us to write Unit tests. Either we pass a pre-selected random numbers for every simulation as an input to test the working of the rest of likelihood test function. Or the 2nd (clever / easy) option was to test each component of the function, separately. We discussed that if, all the parts are individually giving correct results, then they should also be working fine together. Therefore, unit tests for L, M, S and CL tests are based on testing of individual components.

**Comment by wsavran** [comment]
i am working on porting the test cases from csep1 into this pull request

**Pull request merged**